### PR TITLE
Add callback for external view synchronization

### LINKED ISSUE #133990: [Android] Add callback for external view synchronization
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

I've created a plugin for managing external Android views related to https://github.com/flutter/flutter/issues/128198
But it is missing the position synchronization feature. When the external view is moving, its position does not match the frame produced by Flutter.

I was able to implement the synchronization feature but it required exposing some callback from the flutter engine.
This ticket is for merging these changes into Flutter.

### Proposal

The mechanism:
1. At the end of the compositing phase, after updating the view position and before rasterization, send the new position and the timestamp (framework build end time) to the native plugin.
2. At the start of rasterization, send the engine build end time and the current timestamp to the native plugin. (requires adding a new callback into the engine)
3. Using `FlutterImageView `to render the UI. It provides `Image `from `ImageReader`.
4. The `Image `provides `getTimestamp()` which is the nano time when the image is created. Use it to find the rasterization start time -> find the engine build end time -> find the framework build end time -> find the corresponding position. Update the corresponding position and draw the `Image`. (requires adding callback and modifying `FlutterImageView`)

Detail modification:

Engine (PR https://github.com/flutter/engine/pull/45430):
 - Add `FlutterEngine.addOnFrameTimeListener()`, when rasterization starts, it fires the callback that provides buildStartTime, buildEndTime, raster nanoTime for the plugin.
 - Add `FlutterImageView.addOnImageAvailableListener` which exposes `ImageReader.OnImageAvailableListener` callback.
 - FlutterImageView delays the `Image.close()` call until the Image is drawn on screen. Because in practice, calling `Image.close()` too soon will cause miss synchronization event if the `Image` and its position are correct. 
 The image is closed at the `Choreographer.FrameCallback` after the nearest `View.onDraw`. 
 Detail: `ImageReader.acquireLatestImage()` -> first `Choreographer.FrameCallback` -> `View.onDraw()` ->  second `Choreographer.FrameCallback` (Image is closed here).
 - Expose the `FlutterImageView.getPendingImage()`

Result

After:

https://github.com/flutter/flutter/assets/26743089/28fb756d-656b-4d15-a17c-30a8b1f52f82

Before:

https://github.com/flutter/flutter/assets/26743089/49384ba8-9da7-4430-a0eb-4a34317ea252

### PR DESCRIPTION
**Pull request by XuanTung95**
Add callback for implementing external view synchronization.

Fix: https://github.com/flutter/flutter/issues/133990

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/view.dart [hunk 1]
```diff
@@ -7,6 +7,7 @@ import 'dart:ui' as ui show FlutterView, Scene, SceneBuilder, SemanticsUpdate;
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/view.dart:10]: This is a layer violation: The rendering layer cannot import the widget layer.

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 2]
```diff
@@ -756,6 +756,41 @@ mixin SchedulerBinding on BindingBase {
...
+  /// The callbacks are executed in the order in which they have been
+  /// added.
+  ///
+  /// Pre-rendering callbacks cannot be unregistered. They are called exactly once.
+  ///
+  /// See also:
+  ///
+  /// * [addPostFrameCallback], which schedule a callback for the end of
+  ///   this frame.
+  void addPreRenderCallback(VoidCallback callback) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/scheduler/binding.dart:776]: What if you have multiple views? How do I target the rendering of a specific view?

**Comment by goderbauer** [comment]
I don't think that adding this mechanism is a great idea. We should probably study the problem a little more and come up with a better solution.

**Comment by XuanTung95** [comment]
@goderbauer 

I don't understand what is the problem you mentioned.

File: packages/flutter/lib/src/scheduler/binding.dart (hunk 2 above)
Comment (XuanTung95) [reply packages/flutter/lib/src/scheduler/binding.dart:776]: Do you mean multiple views in this? https://github.com/flutter/engine/pull/45512
This feature is not merged into flutter so I don't have the information about the new api/changes need to be made to support it.

File: packages/flutter/lib/src/rendering/view.dart (hunk 1 above)
Comment (XuanTung95) [reply packages/flutter/lib/src/rendering/view.dart]: I changed to flutter/scheduler.dart

File: packages/flutter/lib/src/scheduler/binding.dart (hunk 2 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/scheduler/binding.dart:776]: That PR is just to make engine support rendering to multiple views. The framework's API already supports multiple views. https://api.flutter.dev/flutter/dart-ui/PlatformDispatcher/views.html

**Comment by dkwingsmt** [comment]
Also, it seems you're just adding a `addPreRenderCallback` that your plugin can listen to and send messages to the engine side of the plugin via custom message channels. It feels like an essential feature that should just be shipped with Flutter.

**Comment by XuanTung95** [comment]
@dkwingsmt 

About `addPreRenderCallback`, it is not just a callback, it will get the latest position of my widgets (that's why it must be called after compositing) and send it with the current timestamp to the native side of the plugin. These 2 values will be used to trace back which position corresponds with the Image of ImageReader.

If I implement it in the engine, it must send the position of widgets from dart to the engine, then create a new callback to send it from the engine to the plugin, which is much more complicated and requires more maintenance in the engine.

> Can you show us what the callback for addPreRenderCallback is supposed to be?

It gets the position and current timestamp, sends it using [jni](https://pub.dev/packages/jni). I cannot use MethodChannel because it uses the main looper task, the timing may be late.

**Comment by chunhtai** [comment]
> About addPreRenderCallback, it is not just a callback, it will get the latest position of my widgets (that's why it must be called after compositing) and send it with the current timestamp to the native side of the plugin. These 2 values will be used to trace back which position corresponds with the Image of ImageReader.

Is it possible to create your own binding subclass and override drawFrame to send the message?

**Review by chunhtai — COMMENTED** [review]
see my comment above

**Comment by XuanTung95** [comment]
@chunhtai 

You are right, I can do that. The timing is not perfect but I can work around it.

But it is not user-friendly, it requires additional setup, and may conflict with other plugins which use the same approach.
If it's possible, I still want to merge this.

**Comment by goderbauer** [comment]
Can you hook into the compositing phase by pushing your own layer? The layer tree generally knows where things end up on screen. It shouldn't require a special callback.

**Comment by XuanTung95** [comment]
Good call, I did not think of that, `Layer.addCompositionCallback()` work.
I close this.

**Comment by dkwingsmt** [comment]
Is the engine PR still needed?

**Comment by XuanTung95** [comment]
Yes I still need it

**Pull request closed without merging**